### PR TITLE
Update user functions hubers

### PR DESCRIPTION
**Pull request by cdreisbach**

### CODE REVIEW

File: user/hubers_functions.hpp [hunk 1]
```diff
@@ -10,7 +10,8 @@
 
 #include<constants.h>
 #include<basic_calcs.h>
-#include<NNpoly.h>
+#include "TF1.h"
+#include "TMath.h"
```
Comment (suhlatwork) [inline user/hubers_functions.hpp]: use `<>` for includes.

File: user/hubers_functions.hpp [hunk 2]
```diff
@@ -42,6 +43,36 @@ namespace antok {
 						double* _outAddr;
 				};
 
+				class thetaRICHcut : public Function
+				{
+					public:
+						thetaRICHcut(double* inAddr1, double* inAddr2, double* outAddr)
+										: _inAddr1(inAddr1),
+											_inAddr2(inAddr2),
+											_outAddr(outAddr) {}
```
Comment (suhlatwork) [inline user/hubers_functions.hpp]: indentation

File: user/hubers_functions.hpp [hunk 3]
```diff
@@ -139,31 +175,94 @@ namespace antok {
...
 					public:
-						GetNeuronalBeam(double* xAddr, double* yAddr, double* dxAddr,
-						                double* dyAddr, double* eAddr, TLorentzVector* LVAddr)
-						               :_xAddr(xAddr), _yAddr(yAddr), _dxAddr(dxAddr), _dyAddr(dyAddr),
-						                _eAddr(eAddr), _LVAddr(LVAddr){}
+						GetNeuronalBeam(double* xAddr,
+						                double* yAddr,
+						                double* dxAddr,
+						                std::vector<double> calibration2009,
+						                std::vector<double> calibration2012,
```
Comment (suhlatwork) [inline user/hubers_functions.hpp]: see comment above

File: user/hubers_functions.hpp [hunk 4]
```diff
@@ -139,31 +175,94 @@ namespace antok {
...
 						double* _xAddr;
 						double* _yAddr;
 						double* _dxAddr;
+						std::vector<double> _calibration2009;
+						std::vector<double> _calibration2012;
 						double* _dyAddr;
 						double* _eAddr;
+						int* _yearAddr;
 						TLorentzVector* _LVAddr;
 				};
```
Comment (suhlatwork) [inline user/hubers_functions.hpp]: see comments above for the whole class.

File: user/hubers_functions.hpp [hunk 5]
```diff
@@ -209,6 +308,8 @@ namespace antok {
 						bool operator() () {
 							double fCUT_Z = -50;
 							double fNsigma_theta= 2.5;
+// 							double fCUT_Z0 = 4.5;
+// 							double fCUT_Z1 = 8.5;
```
Comment (suhlatwork) [inline user/hubers_functions.hpp]: commented code

File: user/hubers_functions.hpp [hunk 6]
```diff
@@ -325,7 +426,8 @@ namespace antok {
 								else if((*_method)==1)
 									_resultAddr->push_back( LinearGammaCorrection((*_EAddr)[i]) );
 								else if((*_method)==0)
-									_resultAddr->push_back( PEDepGammaCorrection((*_EAddr)[i], (*_XAddr)[i], (*_YAddr)[i]) );
+									_resultAddr->push_back( (*_EAddr)[i] );
+// 									_resultAddr->push_back( PEDepGammaCorrection((*_EAddr)[i], (*_XAddr)[i], (*_YAddr)[i]) );
```
Comment (suhlatwork) [inline user/hubers_functions.hpp]: * commented code
* method 0 is not doing anything

File: user/hubers_functions.hpp [hunk 7]
```diff
@@ -343,6 +445,99 @@ namespace antok {
 						std::vector<double>* _resultAddr;
 				};
 
+
+
+				//***********************************
+				// Scale Energy of a cluster depending
+				// on energy and position
+				//***********************************
```
Comment (suhlatwork) [inline user/hubers_functions.hpp]: wrong comment for next class

File: user/hubers_functions.hpp [hunk 8]
```diff
@@ -343,6 +445,99 @@ namespace antok {
...
+				class ExtrapNeutral : public Function
+				{
+					public:
+					ExtrapNeutral(double* XAddr, double* YAddr, double* ZAddr,
+												TLorentzVector* piLV, TLorentzVector* beamLV,
+												double* resultAddrX, double* resultAddrY)
```
Comment (suhlatwork) [inline user/hubers_functions.hpp]: indentation

File: user/hubers_functions.hpp [hunk 9]
```diff
@@ -343,6 +445,99 @@ namespace antok {
...
+						*_resultAddrY = gammaLV.Y() / gammaLV.Z() * (*_ZAddr) + (*_YAddr);
+						return 1;
+					}
+
+					private:
+						double *_XAddr;
+						double *_YAddr;
+						double *_ZAddr;
+            TLorentzVector *_piLV;
+            TLorentzVector *_beamLV;
```
Comment (suhlatwork) [inline user/hubers_functions.hpp]: indentation

File: user/hubers_functions.hpp [hunk 10]
```diff
@@ -343,6 +445,99 @@ namespace antok {
...
+						}
+						*_resultAddrX = gammaLV.X() / gammaLV.Z() * (*_ZAddr) + (*_XAddr);
+						*_resultAddrY = gammaLV.Y() / gammaLV.Z() * (*_ZAddr) + (*_YAddr);
+						return 1;
+					}
+
+					private:
+						double *_XAddr;
+						double *_YAddr;
+						double *_ZAddr;
```
Comment (suhlatwork) [inline user/hubers_functions.hpp]: just `Z` is an unlucky choice for the name, something like `Zdiff` / `Zdist` would be better. This should also be changed in the generator.

File: user/hubers_functions.hpp [hunk 11]
```diff
@@ -343,6 +445,99 @@ namespace antok {
...
+
+				//***********************************
+				// Scale Energy of a cluster depending
+				// on energy and position
+				//***********************************
+				class GetClusterPosCor : public Function
+				{
+					public:
+						GetClusterPosCor(std::vector<double>* XAddr, std::vector<double>* YAddr,
+                             std::vector<double>* XicAddr, std::vector<double>* YicAddr,
```
Comment (suhlatwork) [inline user/hubers_functions.hpp]: indentation

File: user/hubers_functions.hpp [hunk 12]
```diff
@@ -343,6 +445,99 @@ namespace antok {
...
+						double* _resultAddrX;
+						double* _resultAddrY;
+				};
+
+
+
+				//***********************************
+				// Scale Energy of a cluster depending
+				// on energy and position
+				//***********************************
```
Comment (suhlatwork) [inline user/hubers_functions.hpp]: wrong comment for next class

File: user/hubers_functions.hpp [hunk 13]
```diff
@@ -343,6 +445,99 @@ namespace antok {
...
+					public:
+						GetClusterPosCor(std::vector<double>* XAddr, std::vector<double>* YAddr,
+                             std::vector<double>* XicAddr, std::vector<double>* YicAddr,
+						                 std::vector<double>* EAddr, int* method, double* threshold,
+						                 std::vector<double>* resultAddrX, std::vector<double>* resultAddrY)
+						                :_XAddr(XAddr),_YAddr(YAddr),_EAddr(EAddr), _XicAddr(XicAddr),_YicAddr(YicAddr),
+						                 _method(method), _threshold(threshold), _resultAddrX(resultAddrX), _resultAddrY(resultAddrY) {
+                              _f==new TF1("lll","[0]*TMath::Sin(x*[1]+[2])",-2,2);
+                              std::cout<<_f<<std::endl;
+//                                _f->SetParameters(0.22,6.28318530717958623/3.83,0);
```
Comment (suhlatwork) [inline user/hubers_functions.hpp]: unused member `_f` and commented code

File: user/hubers_functions.hpp [hunk 14]
```diff
@@ -343,6 +445,99 @@ namespace antok {
...
+						}
+
+					private:
+						std::vector<double> *_XAddr;
+						std::vector<double> *_YAddr;
+						std::vector<double> *_XicAddr;
+						std::vector<double> *_YicAddr;
+						std::vector<double> *_EAddr;
+						int* _method;
+						double* _threshold;
```
Comment (suhlatwork) [inline user/hubers_functions.hpp]: method and threshold are unused

File: user/hubers_functions.hpp [hunk 15]
```diff
@@ -476,7 +671,8 @@ namespace antok {
 						                  _VectorTAddr(VectorTAddr), _VectorEAddr(VectorEAddr),
 						                  _trackX(trackX), _trackY(trackY), _trackT(trackT),
```
Comment (suhlatwork) [inline user/hubers_functions.hpp]: unused variables `track...`

File: user/hubers_functions.hpp [hunk 16]
```diff
@@ -687,7 +883,118 @@ namespace antok {
...
+						virtual ~GetClosestPi0Pi0() {}
+
+						bool operator() () {
+							_outLVAddr1->SetPxPyPzE(9999, 9999, 9999, 99999);
+							*_outMAddr1 = -99;
+							_outLVAddr2->SetPxPyPzE(9999, 9999, 9999, 99999);
+							*_outMAddr2 = -99;
+							double minDist=9e9;
+							int idx2 = 0;
+							if(_eAddr->size() < 4 )
```
Comment (suhlatwork) [inline user/hubers_functions.hpp]: does this method work reliably for anything but four photons?

File: user/hubers_functions.hpp [hunk 17]
```diff
@@ -687,7 +883,118 @@ namespace antok {
...
+								return true;
+							{
+								TVector3 v3a( (*_xAddr)[0] - *_xPVAddr, (*_yAddr)[0] - *_yPVAddr, (*_zAddr)[0] - *_zPVAddr) ;
+								v3a.SetMag((*_eAddr)[0]);
+								TLorentzVector lva(v3a, (*_eAddr)[0]);
+								for(unsigned int j = 1; j < _eAddr->size(); ++j) {
+									if( (*_eAddr)[j] < 2 )
```
Comment (suhlatwork) [inline user/hubers_functions.hpp]: There is no check on the energy of the first cluster, but only of all others.

File: user/hubers_functions.hpp [hunk 18]
```diff
@@ -687,7 +883,118 @@ namespace antok {
...
+										continue;
+									TVector3 v3b( (*_xAddr)[j] - *_xPVAddr, (*_yAddr)[j] - *_yPVAddr, (*_zAddr)[j] - *_zPVAddr) ;
+									v3b.SetMag((*_eAddr)[j]);
+									TLorentzVector lvb(v3b, (*_eAddr)[j]);
+									if(std::fabs((lva + lvb).Mag() - *_selectedMass < minDist)) {
```
Comment (suhlatwork) [inline user/hubers_functions.hpp]: wrong position of closing `)` of fabs

**Comment by tproc12** [comment]
Warum lädst du beide config-files und wo sind die jetzt?

File: user/hubers_functions.hpp [hunk 19]
```diff
@@ -687,7 +883,118 @@ namespace antok {
...
+									TLorentzVector lva(v3a, (*_eAddr)[j]);
+									for(unsigned int jj = j + 1; jj < _eAddr->size(); ++jj){
+										if(jj == idx2)
+											continue;
+										if( (*_eAddr)[jj] < 2 )
+											continue;
+										TVector3 v3b( (*_xAddr)[jj] - *_xPVAddr, (*_yAddr)[jj] - *_yPVAddr, (*_zAddr)[jj] - *_zPVAddr) ;
+										v3b.SetMag((*_eAddr)[jj]);
+										TLorentzVector lvb(v3b, (*_eAddr)[jj]);
+										if(std::fabs((lva + lvb).Mag() - *_selectedMass < minDist)) {
```
Comment (suhlatwork) [inline user/hubers_functions.hpp]: wrong position of closing `)` of fabs

File: user/hubers_functions.hpp [hunk 20]
```diff
@@ -687,7 +883,118 @@ namespace antok {
...
+						double* _zPVAddr;
+						double* _selectedMass;
+						TLorentzVector* _outLVAddr1;
+						double* _outMAddr1;
+						TLorentzVector* _outLVAddr2;
+						double* _outMAddr2;
+				};
+
+				//***********************************
+				//Gets run and spill as one double
```
Comment (suhlatwork) [inline user/hubers_functions.hpp]: why as a double? all numbers are integers, and the numbers are small enough even for signed integers, let alone long longs

File: user/hubers_functions.hpp [hunk 21]
```diff
@@ -687,7 +883,118 @@ namespace antok {
...
+				//***********************************
+				class GetRunSpill : public Function
+				{
+					public:
+						GetRunSpill(int* runAddr, int* spillAddr, double* factorAddr, double* resultAddr)
+						           : _runAddr(runAddr), _spillAddr(spillAddr),
+						             _factorAddr(factorAddr), _resultAddr(resultAddr)
```
Comment (suhlatwork) [inline user/hubers_functions.hpp]: factorAddr is unused

File: src/cut.hpp [hunk 22]
```diff
@@ -200,19 +202,23 @@ namespace antok {
 				  _cutY(cutY),
 				  _X(X),
 				  _Y(Y),
+				  _phi(phi),
 				  _mode(mode) { }
 
 			bool operator() () {
+				double Xrot = (*_X - *_meanX) * std::cos(*_phi) - (*_Y - *_meanY) * std::sin(*_phi);
+				double Yrot = (*_Y - *_meanY) * std::cos(*_phi) + (*_X - *_meanX) * std::sin(*_phi);
```
Comment (tproc12) [inline src/cut.hpp]: What is untypical

Comment (suhlatwork) [inline src/cut.hpp]: its clockwise

File: user/hubers.cxx [hunk 23]
```diff
@@ -805,6 +942,258 @@ antok::Function* antok::user::hubers::generateGetClosestPi0(const YAML::Node& fu
...
 
+//***********************************
+//Gets best pi0pi0 pair
+//gives an LV and the mass
+//***********************************
+antok::Function* antok::user::hubers::generateGetClosestPi0Pi0(const YAML::Node& function, std::vector<std::string>& quantityNames, int index)
+{
+
+	if(quantityNames.size() != 4) {
+		std::cerr<<"Too many names for function \""<<function["Name"]<<" needed four\"."<<std::endl;
```
Comment (tproc12) [inline user/hubers.cxx]: ???

File: user/hubers.cxx [hunk 24]
```diff
@@ -842,8 +1231,10 @@ double antok::user::hubers::PEDepGammaCorrection( double Egamma, double cx, doub
 }
 
 double antok::user::hubers::LinearGammaCorrection( double Egamma ) {
-	static const double fLinEcorr0 = -1.21919;
+	static const double fLinEcorr0 = -1.65+2;//-1.21919-0.8;
 	static const double fLinEcorr1 = 0.033045;
+// 	static const double fLinEcorr0 = -0.859936;
+// 	static const double fLinEcorr1 = 5.76984;
```
Comment (tproc12) [inline user/hubers.cxx]: Should stay in!

File: user/hubers_functions.hpp (hunk 5 above)
Comment (tproc12) [inline user/hubers_functions.hpp]: should stay in!

File: user/hubers_functions.hpp (hunk 6 above)
Comment (tproc12) [inline user/hubers_functions.hpp]: keep it

File: user/hubers_functions.hpp (hunk 14 above)
Comment (tproc12) [inline user/hubers_functions.hpp]: beta code

File: user/hubers_functions.hpp (hunk 16 above)
Comment (tproc12) [inline user/hubers_functions.hpp]: It was not designed for anything else, can be removed completly

File: user/hubers_functions.hpp (hunk 20 above)
Comment (tproc12) [inline user/hubers_functions.hpp]: because plotting of integers vs. doubles was not possible at that time

**Pull request closed without merging**